src/abstract/mina-poseidon.ts

```diff
@@ -0,0 +1,164 @@
+    });
+  });
+
+  if (rev !== undefined && typeof rev !== 'boolean')
+    throw new Error(`Poseidon: invalid param reversePartialPowIdx=${rev}`);
+
+  // if (roundsFull % 2 !== 0) throw new Error(`Poseidon roundsFull is not even: ${roundsFull}`);
+  const rounds = roundsFull + roundsPartial;
+
+  if (!Array.isArray(rc) || rc.length !== rounds)
+    throw new Error('Poseidon: wrong round constants');
+  const roundConstants = rc.map((rc) => {
+    if (!Array.isArray(rc) || rc.length !== t)
+      throw new Error(`Poseidon wrong round constants: ${rc}`);
+    return rc.map((i) => {
+      if (typeof i !== 'bigint' || !Fp.isValid(i))
+        throw new Error(`Poseidon wrong round constant=${i}`);
+      return Fp.create(i);
+    });
+  });
+
+  if (!sboxPower || ![3, 5, 7].includes(sboxPower))
+    throw new Error(`Poseidon wrong sboxPower=${sboxPower}`);
+  const _sboxPower = BigInt(sboxPower);
+  let sboxFn = (n: bigint) => FpPow(Fp, n, _sboxPower);
+  // Unwrapped sbox power for common cases (195->142μs)
+  if (sboxPower === 3) sboxFn = (n: bigint) => Fp.mul(Fp.sqrN(n), n);
+  else if (sboxPower === 5) sboxFn = (n: bigint) => Fp.mul(Fp.sqrN(Fp.sqrN(n)), n);
+
+  return Object.freeze({ ...opts, rounds, sboxFn, roundConstants, mds: _mds });
+}
+
+export function splitConstants(rc: bigint[], t: number) {
+  if (typeof t !== 'number') throw new Error('poseidonSplitConstants: wrong t');
+  if (!Array.isArray(rc) || rc.length % t) throw new Error('poseidonSplitConstants: wrong rc');
+  const res = [];
+  let tmp = [];
+  for (let i = 0; i < rc.length; i++) {
+    tmp.push(rc[i]);
+    if (tmp.length === t) {
+      res.push(tmp);
+      tmp = [];
+    }
+  }
+  return res;
+}
+
+export function poseidon(opts: PoseidonOpts) {
+  const _opts = validateOpts(opts);
+  // const { Fp, mds, roundConstants, rounds, roundsPartial, sboxFn, t } = _opts;
+  const { Fp, mds, roundConstants, rounds, sboxFn, t } = _opts;
+  // const halfRoundsFull = _opts.roundsFull / 2;
+  const partialIdx = _opts.reversePartialPowIdx ? t - 1 : 0;
+  // const poseidonRound = (values: bigint[], isFull: boolean, idx: number) => {
+  //   values = values.map((i, j) => Fp.add(i, roundConstants[idx][j]));
+
+  //   if (isFull) values = values.map((i) => sboxFn(i));
+  //   else values[partialIdx] = sboxFn(values[partialIdx]);
+  //   // Matrix multiplication
+  //   values = mds.map((i) => i.reduce((acc, i, j) => Fp.add(acc, Fp.mulN(i, values[j])), Fp.ZERO));
+  //   return values;
+  // };
+  const poseidonRound = (values: bigint[], isFull: boolean, idx: number) => {
+    if (isFull) values = values.map((i) => sboxFn(i));
+    else values[partialIdx] = sboxFn(values[partialIdx]);
+
+    // Matrix multiplication
+    values = mds.map((i) => i.reduce((acc, i, j) => Fp.add(acc, Fp.mulN(i, values[j])), Fp.ZERO));
+
+    values = values.map((i, j) => Fp.add(i, roundConstants[idx][j]));
+
+    return values;
+  };
+
+  const poseidonHash = function poseidonHash(values: bigint[]): bigint {
```

Comment:
I think in the noble-curves repo this is just supposed to be the permutation, hence the fixed values array size and the lack of rate and capacity parameters. see [here](https://github.com/paulmillr/noble-curves/blob/main/test/_poseidon.helpers.js)  I would probably just keep things as they are to implement the mina-style permutation, and bring the hash and update to a separate file in the style as the referenced test. Mostly it just makes the diff easier to see.